### ISSUE TITLE
`createTag()` of `AttributesTrait` accesses a `private` attribute without a getter

### ISSUE THREAD

**Issue by makoru-hikage** [donquixote/cellbrush#5]
Line 129 of `AttributesTrait` is faulty; it accesses a `private` attribute. The `$attribute` property of `AttributesTrait` used by `Tag` class is private, but it is accessed with the following:

`$tag->attributes = $this->attributes;`

The aforementioned line is Line 129 of `AttributesTrait`.

To summarize what happened: Using the `AttributeTrait`, class `Attributes` accessed the `private $attributes` of `Tag` class. Although both classes `Tag` and `Attributes` use the `AttributesTrait` that has  `private $attributes`, they cannot access such property without doing a 'getter'.

Will think of a PR soon.
[The usage of PHP Traits is quite confusing, isn't it?]

**Comment by donquixote** [comment]
> [The usage of PHP Traits is quite confusing, isn't it?]

Yes, I use them a lot less nowadays. They do not properly encapsulate anything, and make it really hard for an IDE to figure out what's going on. I wish PHP had real multiple inheritance.

**Comment by donquixote** [comment]
Is this not covered by any test, I wonder?

**Comment by donquixote** [comment]
Currently Tag / AttributesTrait are designed as immutable, so all the setters create copies. Nowadays I call such setters not setSomething(), but withSomething(), or withAddedSomething(). E.g. `$car->withColor('blue')`.

Calling such methods always has a tiny little overhead.

Within a constructor or when an object is being freshly created, the immutable aspect is not really needed. I would prefer to not create object copies for no reason.

Maybe we should have a static method `Tag::createWithAttributes($tagName, $attributes)` ? Or add this as an optional parameter to `Tag::__construct()`?

One difficulty with static method constructors is always that they need to call `new static` or `new self`, so they might break in a subclass, if the subclass has a different constructor signature. I already kinda dislike AttributesTrait::create() for this reason.

**Comment by donquixote** [comment]
Or maybe a good design for immutable classes is to have a base class with protected mutable setters, and then a subclass which is immutable because it only calls those setters during construction?

**Comment by donquixote** [comment]
https://twitter.com/brain4lunch/status/786526053640482817 :)

**Comment by makoru-hikage** [comment]
"Or add this as an optional parameter to Tag::__construct()"

This solution seems elegant.

**Comment by donquixote** [comment]
Ok here is a simpler php example.
https://3v4l.org/0HNFH

```php
trait T {
    public function __construct(private int $x = 5) {}
}

class C {
    use T;
    public function foo() {
        $d = new D();
        $d->x = 5;
    }
}

class D {
    use T;
}

(new C)->foo();
```
Indeed we get "Fatal error: Uncaught Error: Cannot access private property D::$x in /in/0HNFH:11"

It seems within this library the `->createTag()` is not really used.
But perhaps other code wants to use it?
